app/models/concerns/tokenable.rb

```diff
@@ -2,12 +2,13 @@ module Tokenable
   extend ActiveSupport::Concern
 
   included do
-    before_create :set_token
+    after_initialize :set_token
   end
 
   protected
 
   def set_token
+    return if self.token
     self.token = loop do
       random_token = SecureRandom.uuid
```

Comment:
uuid if unique by default (it almost impossible to get the same one) no reason to check it in loop